### PR TITLE
adding exists to db interface and other error handling

### PR DESCRIPTION
**Pull request by nitro-neal**
# Overview
* Handle case where `current-list-index >= bitStringLength`
* Only generate the new statusListIndex, and array after doing a read and making sure it does not exist

### CODE REVIEW

File: .github/workflows/steelthread.yml [hunk 1]
```diff
@@ -11,7 +11,7 @@ on:
 
 jobs:
   end2endtest:
-    timeout-minutes: 15
+    timeout-minutes: 20
```
Comment (andresuribe87) [inline .github/workflows/steelthread.yml:14]: Under what scenarios does this take more than 15 min?

File: .github/workflows/steelthread.yml [hunk 2]
```diff
@@ -30,7 +30,12 @@ jobs:
       run: docker-compose -f "build/docker-compose.yml" up -d --build
 
     - name: End To End Test
-      run: mage integration
+      uses: nick-fields/retry@v2
+      with:
+        timeout_minutes: 15
+        max_attempts: 3
+        retry_on: error
+        command: mage integration
```
Comment (andresuribe87) [inline .github/workflows/steelthread.yml:38]: I'm concerned that doing retrys would hide flakiness and real issues we should address. Thoughts?

File: pkg/storage/db_test.go [hunk 3]
```diff
@@ -199,6 +199,37 @@ func TestDBEmptyNamespace(t *testing.T) {
...
+		err = db.Write(context.Background(), namespace, key, nil)
+		assert.NoError(t, err)
+
+		dummyData := []byte("dummy")
+		err = db.Write(context.Background(), namespace, key, dummyData)
+		assert.NoError(t, err)
+
+		exists, err = db.Exists(context.Background(), namespace, key)
+		assert.NoError(t, err)
+		assert.True(t, exists)
```
Comment (andresuribe87) [inline pkg/storage/db_test.go:229]: Would recommend splitting these into three tests.

File: pkg/service/credential/storage.go [hunk 4]
```diff
@@ -80,26 +80,42 @@ type StatusListIndex struct {
...
+	if err != nil {
+		return nil, util.LoggingErrorMsg(err, "exists for statusListIndexesKey")
 	}
 
-	statusListIndexBytes, err := json.Marshal(StatusListIndex{Index: 0})
-	if err != nil {
-		return nil, util.LoggingErrorMsg(err, "could not marshal status list index bytes")
+	if listIndexKeyExists != statusListIndexesKeyExists {
+		return nil, util.LoggingNewError("list index and status list index not in the same state")
 	}
```
Comment (andresuribe87) [inline pkg/service/credential/storage.go:99]: What behavior is this changing? Would recommend adding a test.

File: .github/workflows/steelthread.yml (hunk 1 above)
Comment (nitro-neal) [reply .github/workflows/steelthread.yml:14]: should never happen, it needs the timeout-minutes field to use the action module

File: .github/workflows/steelthread.yml (hunk 2 above)
Comment (nitro-neal) [reply .github/workflows/steelthread.yml:38]: It does fail like 10% of the time, I think it has to do with the nondetermanistic threads test. I have retry logic on our client, but maybe there is some better handling we can do so the failure goes down to like 1% of the time

File: pkg/storage/db_test.go (hunk 3 above)
Comment (nitro-neal) [reply pkg/storage/db_test.go:229]: ok, lots more lines of code though! 🤣

File: .github/workflows/steelthread.yml (hunk 2 above)
Comment (nitro-neal) [reply .github/workflows/steelthread.yml]: removed this, yea lets keep it vanilla

File: pkg/server/router/credential_test.go [hunk 5]
```diff
@@ -275,4 +275,72 @@ func TestCredentialRouter(t *testing.T) {
 		assert.NotEmpty(tt, credStatusMapTwo["statusListIndex"])
 
 	})
+
+	t.Run("Credential Status List Test On Create", func(tt *testing.T) {
```
Comment (andresuribe87) [inline pkg/server/router/credential_test.go:279]: It's unclear what **behavior** this test is exercising. Would recommend following this [naming best practices](https://learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-best-practices#naming-your-tests) for this test as well. Perhaps worth also splitting it up.

File: pkg/storage/db_test.go [hunk 6]
```diff
@@ -199,6 +199,72 @@ func TestDBEmptyNamespace(t *testing.T) {
...
 
+func TestDBExistsNoNamespaceKey(t *testing.T) {
+	for _, dbImpl := range getDBImplementations(t) {
+		db := dbImpl
+
+		exists, err := db.Exists(context.Background(), "dnenamespace", "dnekey")
+		assert.NoError(t, err)
+		assert.False(t, exists)
+	}
+}
```
Comment (andresuribe87) [inline pkg/storage/db_test.go:210]: This is awesome! Small suggestion to also include the expected behavior in the test, so reader can quickly verify that what you're asserting it correct.
```go
func TestDBExists_FalseWhenNoNamespaceKey(t *testing.T) {
	for _, dbImpl := range getDBImplementations(t) {
		db := dbImpl

		exists, err := db.Exists(context.Background(), "dnenamespace", "dnekey")
		assert.NoError(t, err)
		assert.False(t, exists)
	}
}
```

File: pkg/storage/db_test.go [hunk 7]
```diff
@@ -199,6 +199,72 @@ func TestDBEmptyNamespace(t *testing.T) {
...
+
+func TestDBExistsNoKey(t *testing.T) {
```
Comment (andresuribe87) [inline pkg/storage/db_test.go:212]:
```suggestion
func TestDBExists_FalseWhenKeyIsAbsent(t *testing.T) {
```

File: pkg/storage/db_test.go [hunk 8]
```diff
@@ -199,6 +199,72 @@ func TestDBEmptyNamespace(t *testing.T) {
...
+		assert.NoError(t, err)
+
+		exists, err := db.Exists(context.Background(), namespace, "otherdnekey")
+		assert.NoError(t, err)
+		assert.False(t, exists)
+
+	}
+}
+
+func TestDBExistsNilValue(t *testing.T) {
```
Comment (andresuribe87) [inline pkg/storage/db_test.go:228]: You get the point :)

File: pkg/storage/db_test.go [hunk 9]
```diff
@@ -199,6 +199,72 @@ func TestDBEmptyNamespace(t *testing.T) {
...
+		assert.True(t, exists)
+
+		err = db.Delete(context.Background(), namespace, key)
+		assert.NoError(t, err)
+
+		exists, err = db.Exists(context.Background(), namespace, key)
+		assert.NoError(t, err)
+		assert.False(t, exists)
+	}
+}
```
Comment (andresuribe87) [inline pkg/storage/db_test.go:266]: Would recommend naming `TestDBExists_FalseAfterDeletingKey`

File: pkg/service/credential/storage.go [hunk 10]
```diff
@@ -80,26 +80,42 @@ type StatusListIndex struct {
...
 
-	if err := db.Write(context.Background(), statusListIndexNamespace, currentListIndexKey, statusListIndexBytes); err != nil {
-		return nil, util.LoggingErrorMsg(err, "problem writing current list index to db")
+	if !listIndexKeyExists && !statusListIndexesKeyExists {
```
Comment (decentralgabe) [inline pkg/service/credential/storage.go:100]: nit: consider a utility variable to help make this more readable

like `indexKeysDontExist := !listIndexKeyExists && !statusListIndexesKeyExists`

**Review by decentralgabe — APPROVED** [review]

**Pull request merged**